eth_p2p/blockchain_sync.nim

```diff
@@ -72,25 +72,31 @@ proc availableWorkItem(ctx: SyncContext): int =
   ctx.workQueue[result] = WantedBlocks(startIndex: nextRequestedBlock, numBlocks: numBlocks.uint, state: Initial)
 
 proc persistWorkItem(ctx: SyncContext, wi: var WantedBlocks) =
-  ctx.chain.persistBlocks(wi.headers, wi.bodies)
+  case ctx.chain.persistBlocks(wi.headers, wi.bodies)
+  of ValidationResult.OK:
+    ctx.finalizedBlock = wi.endIndex
+    wi.state = Persisted
+  of ValidationResult.RecoverableValidationError:
+    wi.state = Initial
+  of ValidationResult.UnrecoverableValidationError:
```

Comment:
For unrecoverable errors, just go with `doAssert`. This will throw an exception that will eventually lead to Nimbus being terminated. The user will have to restart the client, which is considered a better alternative than operating with a potentially corrupted program state (with us being financial software, it's better to be safe than sorry).

Reply:
If you read my initial explanation, you'll see there is no contradiction:
https://github.com/status-im/nim-eth-p2p/pull/55#discussion_r241794489

 `asserts` are used for non-recoverable errors (detected bugs) and they should terminate the program in most cases (after capturing as much debugging information as possible). Recoverable errors are things that the calling code is required to handle (because they can arise under normal operation of the program). It's bad style to mix exception and error code only for the recoverable errors.

Reply:
> Asserts are used for non-recoverable errors (detected bugs)

Apparently, they are also "merely mocks of proper error handling, and replacing them with doAsserts doesn't make the code any more correct": https://github.com/status-im/nimbus/pull/199#issuecomment-447334915

> they should terminate the program in most cases

In this case, RPC debugging might want to intercept this error and replay the transactions in order to extract more debugging data. Either exceptions or error codes will work for this.